RFC: file: specifier changes

iarna opened the pull request:
Easy reading link for the new [specification](https://github.com/npm/npm/blob/link-specifier/doc/spec/file-specifiers.md). This also introduces written specifications for npm features, something we intend to extend out to all of npm's activity.

# The Problem We're Solving

File type dependency specifiers are confusing. How they interact across npm's many varied commands (`npm install --save`, `npm update`, `npm outdated`, `npm shrinkwrap`) has never been defined in one place. Most of the behaviors "happened by default". That is, the minimum implementation was done to make them install ok and the rest was left as a side effect.

Currently `npm update` and `npm outdated` simply don't do anything for local dependencies. If you've updated the source you have to manually reinstall to get a copy of it. How `npm shrinkwrap` saves local dependencies to `npm-shrinkwrap.json` and how it resolves them has varied over time as well.

## Root Causes

When local dependencies were added, we didn't have any process around defining behavior or ensuring that all use cases were specified.

## Proposal

In addition to actually specifying behavior (often simply writing down what
things currently do), we propose in important breaking change:

* `file:`-type specifiers that refer to directories will be soft deprecated, their behavior being identical to the new `link:` specifier and their existence becoming a footnote in the documentation.
* A new specifer type, `link:`, will be introduced for linking local directories.  For the duration of `npm@5, `file:` specifiers that refer to directories will be treated identically to `link:` specifiers.
* `link:` specifiers will result in a symlink or junction made from the specifier path into your `node_modules`.  On Windows try a junction and if that fails, try a symlink.  If both fail, the error from the junction should be used.

This RFC essentially brings [linklocal](https://github.com/timoxley/linklocal)'s bevior into core.

## RISKS

* Diagnostic information changes need to be handle delicately in order to not increase user confusion.
* Some users may be unhappy with the changes to `file:` semantics.

iarna committed:
spec: Describe local specifiers

This introduces changes to how `file:` specifiers work.  Specifically, they
now create symlinks.

timoxley commented:
Nice! Why new `link:` vs just using `file:`?

iarna commented:
The reason for introducing the new specifier are:

1. Have a specifier that immediately tells someone reading a `package.json` what it will do.
2. Not have fights over what a valid `file:` URL looks like.

Originally there was a plan to actually hard deprecate the `file:` specifier and remove them, but we decided that was pointlessly disruptive. Instead `file:` will be an alias for `link:`.

timoxley commented:
@iarna Sounds reasonable 👍 

Have you considered the impact of having different module resolution paths when using symlinks? 
I currently only use symlinks in development, while in production I let the current `file:` behaviour do its thing. Without the symlinks npm is able to move many shared packages to the top level, decreasing overall install time. This could have significant impact on deploy times if local dependencies share a lot of binary dependencies.

iarna commented:
@timoxley It's a fair question. As it's specced right now it would do that. Maybe, if the linked module is inside the top level install root we should flatten its deps down, instead of keeping them inside the linked module. The module loader would still be able to pick them up and it'd not increase disk usage.

timoxley commented:
> if the linked module is inside the top level install root we should flatten its deps down

@iarna that sounds good. Being aware of where symlinks resolve to perhaps could be considered a good general improvement to the dedupe/flattening algorithm.

thefourtheye commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,159 @@
+* Attempting to install a specifer that has a windows drive letter will
+  produce an error on non-Windows systems.
+
+* A valid `file:` specifier points is:
```
Did you mean "A valid `file:` specifier also is" like in the next item? I am finding it difficult to understand this line.

legodude17 replied:
I would say it should be "Also, a valid `file:` specifier is:"

mantoni commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,159 @@
+
+Historically, these ambiguous specifiers were also allowed in the
+`package.json`.  Starting in `npm@5` using an ambiguous specifier in your
+shrinkwrap will be depricated and will warn.  In `npm@6` it will be an
```
Typo: depricated -> deprecated

mantoni commented:
This is great! I would like to know if you considered supporting the current `npm ln` use-case? I'm thinking of `link:module-name` to resolve to the "global" link which in turn links to the actual module directory. Contrary to the relative link paths, this wouldn't enforce a specific directory structure. What do you think?

iarna committed:
install: Replace post install module list with summary report

legodude17 commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,159 @@
+* A valid `file:` specifier points is:
+  * a valid package file. That is, a `.tar`, `.tar.gz` or `.tgz` containing
+  `<dir>/package.json`.
+  * OR, a directory that contains a `package.json`
```
Should that 'OR' be removed or should all of these start with 'OR'? Consistency.

iarna replied:
There are only two of them and the OR separates the two. So I'm not sure what you mean by "all of these".

legodude17 commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,159 @@
+  * a valid package file. That is, a `.tar`, `.tar.gz` or `.tgz` containing
+  `<dir>/package.json`.
+  * OR, a directory that contains a `package.json`
+* And a valid `file:` specifier also is:
```
Shouldn't this not be a bullet? Or just remove it.

iarna replied:
It should. It's part of the top level list.

legodude17 commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,159 @@
+The `preinstall` for file-type specifiers MUST be run AFTER the
+`finalize` phase as the symlink may be a relative path reaching outside the
+current project root and a symlink that resolves in `.staging` won't resolve
+in the package's final resting place.
```
Would the `preinstall` for a package depended on via `link:` be run from it's original folder, or somewhere else?

iarna replied:
Basically this is mean trying to not screw people who are using `pwd` (which would be relative to where your package root's node_modules).

Folks who are using `cwd` would always get the same answer (the destination of the link).

legodude17 commented on doc/spec/file-and-link-specifiers.md:
````diff
@@ -0,0 +1,159 @@
+```
+example-package@1.0.0 /path/to/example-package
++-- a -> Copied from: link:../a
+```
````
Again, is this repetition intentional?

legodude17 commented on doc/spec/file-and-link-specifiers.md:
````diff
@@ -0,0 +1,159 @@
+```
+example-package@1.0.0 /path/to/example-package
++-- a -> link:../a
+```
````
Why are there two parts with the same thing? Is that intentional?

legodude17 reviewed (commented):
This seems like it would really help a lot of people. 🎆

iarna commented:
So after some discussion in air between @zkat and I, we're of a consensus that we should go ahead with just `file:` and not add a new specifier. I'll update the RFC as appropriate.

zkat commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,156 @@
+#### File type specifers pointing at directories
+
+File-type specifiers that point at directories will necessarily not do
+anything for `fetch` and `extract` phases.
```
How do you feel about pacote doing this automagically on `pacote.extract`?

iarna replied:
Hrm. I don't think pacote has quite enough information currently to resolve this sort of thing. The specifier and the target destination aren't enough. You also need to know the location of the module that required it, because the specifier is relative to THAT.

zkat replied:
Can/should we add this to `realize-package-specifier`? We use `where` there already.

iarna replied:
Maybe, I mean, we might actually already be resolving come to think. It just requires some end-to-end care.

zkat commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,156 @@
+    `file:///foo/bar` reference the same package.
+* … or a relative path (eg `../path/to/thing`, `path\to\subdir`).  Leading
+  slashes on a file specifier will be removed, that is 'file://../foo/bar`
+  references the same package as same as `file:../foo/bar`.  The latter is
```
Can you add a quick clarification that `file://foo/bar` is considered relative, too? At least that's what I assume from reading this. It's a weird, sandwichy corner case, but it's probably worth actually _specifying_:

* `file:///foo` -> Absolute `/foo`
* `file://foo` -> Relative, same as `./foo`
* `file:/foo` -> Absolute `/foo`

iarna replied:
Number of leading slashes changes nothing:

file:///foo -> /foo
file://foo -> /foo
file:/foo -> /foo
With no leading slashes, relative paths are evident:

file:foo -> ./foo
file:../ -> ../foo

I'm proposing that if you have ../ as your first path element then we ignore any leading slashes:

file:/../foo -> ../foo
file://../foo -> ../foo
file:///../foo -> ../foo

I think this would eliminate whole classes of likely errors, particularly around users who are trying to make these things work like URLs.

zkat commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,156 @@
+* Attempting to install a specifer that has a windows drive letter will
+  produce an error on non-Windows systems.
+* A valid `file:` specifier points is:
+  * a valid package file. That is, a `.tar`, `.tar.gz` or `.tgz` containing
```
We should make sure the error message from this is VERY CLEAR that we determine tarballness based on file extension. I know we talked about this a lot, but I still want to avoid linuxy/unixy folks pulling their hair out if they, say, try to install a tarball they downloaded from a webservice that does not specify a filename and they end up with an unsuffixed filename that is still technically a tarball.

Or, for example, if someone tries to `npm install ~/.npm/.pacote/content/deadbeef`, which do not have `.tgz` suffixes.

zkat commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,156 @@
+
+#### File type specifiers pointing at tarballs
+
+File-type specifiers pointing at a `.tgz` or `.tar.gz or `.tar` file will
```
missing backtick (`) here after .tar.gz

legodude17 commented on doc/spec/file-and-link-specifiers.md:
````diff
@@ -0,0 +1,156 @@
+
+```
+example-package@1.0.0 /path/to/example-package
++-- a -> file:../a
````
Still, is this repetition intentional?

zkat replied:
what repetition? directories and package names can be different: `a -> file:../b` is perfectly valid

iarna replied:
One of these is with unicode enabled, one without. I wanted to specify both.

legodude17 replied:
Oh. Sorry. I did't notice that.

legodude17 commented on doc/spec/file-and-link-specifiers.md:
````diff
@@ -0,0 +1,156 @@
+
+```
+Package  Current  Wanted  Latest  Location
+a        MISSING   LOCAL   LOCAL  example-package
````
I would suggest that it should also log where it is looking for the package in.

zkat replied:
oh god `Location` is such a horrible, confusing name when we factor this in. I really bloody wish we had a different name for that column.

zkat replied:
Additionally, I wish we had somewhere to put more information about the package getting installed, such as `from` or something, because the package name by itself doesn't give us much. But that's something for a separate RFC.

iarna replied:
Yup, I'd love to change that too

iarna replied:
As zkat says below, changing outdated to be more useful is really a separate RFC.

legodude17 replied:
Ok, cool. Sounds like a really good thing.

legodude17 replied:
Wait, I am confused. So `package` is the name, and `Location` is the path on disk? Or is it the path to it through the dependency chain? Or is it the package that depended on this?

zkat replied:
@legodude17 `npm help outdated` will answer your questions here.

legodude17 replied:
@zkat thanks for the tip. You are right that `Location` is a very confusing name for that. Maybe `Parent`?

zkat replied:
@legodude17 we'll talk about that once there's an RFC for it. Talking about it in here is bound to get lost in oblivion (and is offtopic for this PR)

legodude17 replied:
Sounds like a good idea. Do you have a estimated time for this?

iarna replied:
Nope, not part of the current schedule. I have on my todo-list writing up an up-to-date "state of NPM" blog post to bring everyone up-to-date on our goals and plans and expectations.

zkat commented on doc/spec/file-and-link-specifiers.md:
```diff
@@ -0,0 +1,156 @@
+dependencies of the linked package will be hoisted to the top level as usual.
+
+If the module is outside the package root then dependencies will be installed inside
+the linked module's `node_modules` folder.
```
Is this going to be done as a separate npm run within that project folder (possibly a subprocess?) or will we build a single ideal tree with some special semantics, and do it in a single npm run? This seems trickier to me than what this single sentence describes. 🤔

iarna replied:
Nah, single thing, it doesn't even require much in the way of special semantics. It's basically the same logic as how we define "global" mode currently.

zkat reviewed (changes requested):
pretty excited about this. Thanks for writing the spec up <3

mlucool commented:
Using symlinks with node, requires a flag set for it to work correctly: https://github.com/nodejs/node/pull/8749#pullrequestreview-1388063. Is the intent to change how node handles symlinks also?

One nice feature about how file works today is for peer dependencies. Due to the copy instead of a symlink, peers with global singletons work as expected. When this moves to symlinks only, the resolution will make certain cases impossible to resolve. Example:
* Module A is a react component and has react as a `peerDependency` and `devDependency`
* While working on A, I want to test this locally in Module B , so I use `file:/path/to/a`.
  * Today it will do the right thing and share the global version of react as it is copied. The dependency resolution won't include a second react.
  * If instead we did a symlink, this would not work because React is installed in in A's `node_modules` and would not be the same instance as in as B's `node_modules`. 

Let me know if that example needs more clarification or if I have misunderstood RFC.

zkat commented:
@mlucool my understanding of this case is that it would continue to work so long as Module A is anywhere inside the directory structure for B when you link it.

That is:

```
module-b
\
 | - module-a
 |   \
 |     node_modules
 \
  node_modules
  \
    react
```

`module-a` will have its `react` hoisted to the top. It's worth being more explicit about how we handle peerDeps in this situation. I hopefully didn't misunderstand the RFC in this case.

@iarna it might be worth specifying the logic for `peerDependencies` and `devDependencies` for these linked submodules in the RFC, cause I don't think I'm that clear on it rn.

mlucool commented:
@zkat In my case A is outside of B always. We do not have a monorepo. Still [this](https://github.com/npm/npm/blob/43db0527310770647f3e9d0b0d76dce53e767de6/doc/spec/file-and-link-specifiers.md#installation) seems to imply as long as it is note `tar` it'll use symlinks.

My current hack (which works quite well) is to create a symlink like util that does:
1. Uninstall A
2. Install via file:/path/to/a
3. Start a file watcher for all files in A's `package.json:files`
4. Start a file watcher for A's package.json to start from 1 again

Now for any change I make it appears nearly instantly. Clearly this is a hack, but it maintains that I can separate components across repos and not worry about dependency issues.

zkat commented:
@mlucool if you're not going to use a monorepo structure, you could literally just run `npm pack` and _then_ `npm install file://path/to/module-a/module-a-1.2.3.tgz`, which will have the same behavior as `npm i file://path/to/module-a` currently does. We're not changing local tarballs. It's a single extra step, but also something the CLI was already doing (it had to, in order to install the package)

zkat commented:
and yes, we will _always_ create symlinks for _directory_ specs. The difference is that we hoist deps normally _iff_ the target package is in a subdir of the current package.

mlucool commented:
@zkat I could do that to keep my hack working. I am still a little unsure about how this hoisting will play out because you can imagine something like:
```
examples
\
  example1
    \ 
     package.json
components
\
  component1
   \ 
    package.json
```
If I did a `file:` in example1, this would not work as far as I can tell now (assuming this is still a React example). I *believe* this paradigm is relatively common also.

Any comments on the node flag required for this to work?

P.S. I am a BIG fan of better support for symlinks, per my commit to node.

zkat commented:
@mlucool It's worth noting that it's tremendously unlikely for us at npm to try to change module resolution, including symlink stuff, ourselves. afaik, there's no intent for us to actually go and push for a change on that end in order to land this.

More likely, I think, is discussing the possibility of taking that flag into account when deciding how to build the hoisted tree -- some users will want that symlink preservation, some won't. I'll defer to @iarna about whether she thinks this is something worth exploring, with a note that any current hacks around `directory` dependencies can continue to exist with that `npm pack` thing I mentioned above.

@iarna does it make sense to you to read `process.env.NODE_PRESERVE_SYMLINKS` from the env and hoisting across symlink boundaries if we find that flag?

mlucool commented:
IMO if something is outside of a module you should never touch its contents.

iarna commented:
@mlucool As @zkat says, changing the default Node.js module loader behavior isn't on the table, which is why 'preserve-symlinks' landed as an option.

Thank you for bringing this up! What you described isn't a use-case we had discussed previously.

I think having it change install behavior based on `NODE_PRESERVE_SYMLINKS` is actually mandatory. `npm` as part of installation has to model how Node.js would load the package to correctly determine where to put its dependencies. This is a detail we've thus far not had to deal with because `npm install` has treated symlinks as opaque things it can't know details of.

So yeah, if you have `NODE_PRESERVE_SYMLINKS` then npm will install transitive dependencies of linked dependencies in your local project, not in the symlinked module. That is, it will work the same way it works when the symlinked module is in a subdirectory of your project.

iarna committed:
test: Spec tests for local specifiers

benlangfeld commented:
@iarna This looks great, and would resolve a large number of headaches for my team using a monorepo. What can I do to help this make progress?

zkat commented:
@benlangfeld not much right now. We're working towards an implementation now. You can track #16244, which is what this is going to be included in.

benlangfeld commented:
Ok, thank you @zkat. If there's anything you can make use of warm bodies for to help with this, you know where we are :)

iarna commented:
@benlangfeld Yeah, it's my current task. =) The actual implementation isn't too tricky, most of the work has been in tests.

iarna committed:
install: Print warnings before summary

iarna committed:
tests: Update tests to support new output formats

iarna committed:
dedupe: find-dupes should output _something_

npm find-dupes is now an alias for `npm dedupe --dry-run --parseable`

iarna committed:
build: Only output symlink info when doing human output

iarna committed:
ls: Remove blank lines from `npm ls --parseable`

zkat committed:
test: fixing standard fails

iarna committed:
install/deps: Remove needlessly async calls to npa

Credit: @iarna

zkat committed:
fix(save): calculate the version to save for new registry deps (#16414)

iarna committed:
install: Fetch as much registry metadata we can at once (#16422)

iarna committed:
decompose-actions: Remove npat test actions

iarna committed:
filespecs: Implement new file: specifier behavior

PR-URL: https://github.com/npm/npm/pull/15900
Credit: @iarna

actions: Allow actions to return promises

extract: Eliminate extra callback

finalize: Rewrite finalize as promises

update-linked: Remove update-linked action

install: Remove obsolete invalid action filtering

fetch-package-metadata: Error on installing windows paths on non-windows systems

fetch-package-metadata: Read in modules installed inside of new links

finalize: Act on realpaths because we may be installing inside a link prior to the symlink being made

finalize: Create symlinks of directory deps

deps: Resolve ambiguity for file specifiers in the traditional way

deps: Set link and realpath properties for directory deps

inflate-bundled: Distinguish published bundles from lined modules

node: Add new fromLink attribute to track sourced-to-symlinks deps

diff-trees: Don't try to install deps that are already inside a link

fetch-package-metadata: Improve error messages for link failures

install: Run preinstall lifecycle after finalize

finalize: Start reading package.json files here

decompose-actions: Don't fetch or extract links

deps: Determine if a link matches spec based on where it points

deps: Compute correct save specifier for file spec

deps: Set the isInLink property on new children

deps: When finding the install location, don't walk up out of a symlink unless PRESERVE_SYMLINKS is on

diff-trees: Stop setting isInLink, this is now a first class property

diff-trees: Only exclude children of links from adding if they were already there behind the symlink

inflate-bundled: We are using isInLink now not fromLink

inflate-bundled: realpaths should be built on realpaths

node: Eliminate fromLink as a thing

node: Fill in values for inLink, isInLink & fromBundle

save: When updating a lock/shrinkwrap don't read the damn tree again

shrinkwrap: Fill in version per the new shrinkwrap spec

install: Make _inBundle purely a debugging artifact

inflate-shrinkwrap: Fix how bundle deps are inflated

save: Detect dependency type when saving

recalculateMetadata: Rewrite as simpler and synchronous

prune: Work off computed metadata

refreshPackageJson: Copy all of the in memory version over the on-disk version except empty items

install: Reduce normalize tree function to minimum

realize-package-specifier: Bring into closer alignment with spec

read-shrinkwrap: Use child.isTop not !child.parent

copy-tree: Don't crash of requires or requiredBy are missing

copy-tree: Initialize copied nodes

dedupe: use getRequested instead of child.package._requested

install: recompute the pkg relationships after finalize

Neccessary to align behavior with shrinkwraps where relationship information
isn't available until we read the module off disk.

install: Prune the tree after building it

install: Only save deps if we made changes

inflate-shrinkwrap: Better metadata while inflating

zkat committed:
pacote@2.7.4

zkat committed:
fix(tests): getting test suite working again

zkat committed:
feat(install): use pacote for package downloads

zkat committed:
fix(shrinkwrap): make sure resolved is there

zkat committed:
bring stuff closer to old version

zkat committed:
refactor(shrinkwrap): refactor and promisify inflate-shrinkwrap.js

zkat committed:
fix(shrinkwrap): include and use bundle information in shrinkwrap

zkat committed:
fix(noise): stop the gentlyRm spam omg

zkat committed:
fix(tests): get.js is no longer used for package downloads and is not cached

zkat committed:
fix(tests): update retry-on-stale-cache to use --prefer-offline

zkat committed:
fix(test): updated splat-with-only-prerelease-latest test

zkat committed:
feat(opts): save installs by default

BREAKING CHANGE: you need to use --no-save to skip saves now;

zkat committed:
feat(logger): added `notice` loglevel

zkat committed:
feat(lockfile): add package-lock.json support

This implements https://github.com/npm/npm/pull/16441 as well as the auto-save behavior it needs in order to be kept up to date.

zkat committed:
FLOATING PATCH: directly patch fstream-npm

remove this patch once https://github.com/npm/fstream-npm/pull/27 lands

zkat committed:
feat(clean): hide npm cache clean behind --force

zkat committed:
fix(cache): reject any attempt to remove individual packages

zkat committed:
meta: dogfood package-lock.json

zkat committed:
fix(extract): cut down on EBUNDLEOVERRIDE spam

zkat committed:
fix(shrinkwrap): should still work if lockfile has no deps

zkat committed:
refactor(extract): covert extract.js to Promises

zkat committed:
feat(shrinkwrap): always shrinkwrap devDeps, even if only devDeps are saved

iarna committed:
fix(shrinkwrap): force installs to use the shrinkwrap tree

zkat committed:
shrinkwrap: missed more spots about shrinkwrap

zkat committed:
cleanup: fix standard bits

zkat committed:
tests: Expect package-locks

iarna committed:
error-message: Stop gaslighting our users with "this is probably your fault"

iarna committed:
common-tap: Make environment objects immutable

zkat committed:
test: disable shrinkwrap for self-install test

zkat committed:
test: npm it test was clobbering package-lock.json

zkat committed:
fix(shrinkwrap): write lockfile to the tree root

zkat committed:
fix(save): pay attention to save type

zkat committed:
save: Always update even if nothing changed

zkat committed:
pacote@2.7.5

zkat committed:
fix(extract): stop trusting package.json in the tarball

zkat committed:
feat(git): git repos with prepare scripts are now built on install

zkat committed:
fix(version): properly detect whether lockfile or shrinkwrap

zkat committed:
pacote@2.7.6

zkat committed:
npmlog@4.1.0

zkat committed:
feat(log): moved new loglevels to npmlog proper

zkat committed:
feat(indent): package.json and package locks preserve indentaton

zkat committed:
detect-indent@5.0.0

zkat committed:
fix(shrinkwrap): npm shrinkwrap should create a shrinkwrap by default, not a pkglock

zkat committed:
fix(tests): shrinkwrap always includes the full tree

zkat committed:
fix(test): dev deps are included in shrinkwraps by default now

zkat committed:
pacote@2.7.7

zkat committed:
fix(git): integrity hash was failing sometimes

zkat committed:
fix(tests): removing obsolete shrinwkrap tests

zkat committed:
fix(deps): make sure to catch errors here

zkat committed:
fix(shrinkwrap): return shrinkwrap data even if renaming

zkat committed:
fix(tests): update override-bundled

zkat committed:
fix(unpublish): stop clearing caches for now

zkat committed:
tests: move pending tests out of the way

zkat committed:
fixup: standard fixes

zkat committed:
test: no need to call shrinkwrap

zkat committed:
deps: upgrade transitive deps

zkat committed:
fix(bundles): use versions for bundleDeps

zkat committed:
fix(update-package-json): crasher when devDeps wasn't set at all

iarna committed:
inflate-shrinkwrap: Remove last of support for partial shrinkwraps

iarna committed:
install: Don't show removed or failed modules in debugging trees

iarna committed:
install: Still save changes when node_modules started out empty

zkat committed:
meta: regenerate package lock

iarna committed:
inflate: Integrity matches mean things are the same

iarna committed:
shrinkwrap: Only save _resolved to the shrinkwrap if there is one

iarna committed:
inflate-shrinkwrap: Don't recompute shrinkwrap values

If everything we know about a dep came from the shrinkwrap file then when it
comes to write it back out to the shrinkwrap just use what we originally
had.

iarna committed:
inflate-shrinkwrap: Allow matching of sw.version and resolved as that's legit now

iarna committed:
shrinkwrap: When saving non-registry deps, handle missing _resolved

Fallback to using a value from _requested or `getRequested`. Previously
it would fall through to using the version number, which is never
correct for non-registry dependencies.

iarna committed:
shrinkwrap: Don't include integrity for git deps

Git deps should be resolved to a specific git ref, which guarantees
a specific version as thoroughly as an integtrity value.

Currently integrity values are _also_ used as content addresses for cacache
and git sources don't have those.

iarna committed:
pacote@2.7.9

zkat committed:
node-gyp@3.6.1

iarna committed:
finalize: If we go to create a link, its ok if its already there

Specifically, if we fail to create the link then it's ok of the link already
exists as long as it points at the same place.

iarna committed:
decompose-actions: Don't finalize bundled symlinks

iarna committed:
get-requested: Resolve using the correct root path

iarna committed:
inflate-bundled: Leave realpaths alone for inflated links

iarna committed:
inflate-shrinkwrap: Check link equivalency for shrinkwraped file specs

iarna committed:
common-tap: Make common.npm support returning promsies

iarna committed:
test: rewrite bundled-dependencies-nonarray to not use link deps

iarna committed:
fetch-package-metadata: Reduce spam to error log on errors

iarna committed:
node: Make sure we _really_ only add new keys to the top level

iarna committed:
npm-registry-couchapp@2.6.13

iarna committed:
test: tree-style update to not use links

iarna committed:
test: uninstall-save now passes

iarna committed:
tests: Categorize failing-for-npm5 tests

iarna committed:
test: Move back passing tests

iarna committed:
travis: ditch npm5 specific stuff

Also prune after install, instead of before.

iarna committed:
test: fix spec-local-specifiers under nyc

iarna committed:
doctor: Make access checks v4 compatible

iarna committed:
version: Fix on v4

iarna committed:
tests: Filter run-script env

zkat committed:
meta: standard fixes

iarna committed:
test: tagged-version-matching

iarna committed:
test: shrinkwrap-version-match

iarna committed:
error-message: Remove last remaining "not a bug" error message

iarna committed:
shrinkwrap: Make API always return the shrinkwrap

iarna committed:
test: tests that pass w/ shrinkwrap API fix

iarna committed:
finalize: Don't nuke symlink destinations pls

iarna committed:
finalize: Allow symlinking over paths much like we allow finalizing over paths

iarna committed:
test: shrinkwrap-nested

zkat committed:
fix(install): tree is null sometimes

when loading up globals, metadata is fetched before
we have an actual tree. This guard is fine here.

iarna committed:
test: prune

iarna committed:
finalize: call next when rolling back symlinks

zkat committed:
fix(pkg-lock): don't crash if code falls through to semver w/o a version

iarna committed:
test: ls-l-depth-0

iarna committed:
test: ls-depth-cli

iarna committed:
test: ls-depth-unmet

iarna committed:
test: shrinkwrap-default-arg-ver, feature removed

iarna committed:
test: shrinkwrap-local-dependency

iarna committed:
test: install-report-just-installed

zkat committed:
feat(shrinkwrap): support registry-switching

zkat committed:
fix(shrinkwrap): reuse sw resolved for diff-registry children

zkat committed:
fix(save): write a package-lock even if no changes made

zkat committed:
fix(pkg-lock): refetch metadata for unfamiliar pkgs

zkat committed:
finalize: standard fixes

iarna committed:
test: override-bundled

iarna committed:
test: categorize

iarna committed:
test: gently-rm-overeager

iarna committed:
test: install-noargs-dev

zkat committed:
test: fix dedupe-scoped test

zkat committed:
test: shrinkwrap-save-dev-with-existing-deps

zkat committed:
fix(link): a package is a link if any of its ancestors is a link

zkat committed:
test: shared-linked

zkat committed:
test: config-meta

zkat committed:
pacote@2.7.10

zkat committed:
fix(shrinkwrap): promisify to fix dropped errors

zkat committed:
pacote@2.7.11

zkat committed:
actions: converted actions.js to use mainly promises

zkat committed:
finalize: fix some dropped error issues

zkat committed:
actions: bump concurrent action count

iarna committed:
test: categorize add-remote-git-get-resolved

zkat committed:
fstream-npm@1.2.1

zkat committed:
standard: clean up test directories

zkat committed:
write-file-atomic@2.1.0

adds fsync option

zkat committed:
update-package-json: disable fsync on file write

zkat committed:
fix(auth): resolve _resolved with the package name to get auth working again

zkat committed:
pacote: pass the user-agent through

zkat committed:
pacote@2.7.12

iarna committed:
git-prepare: Fix child process npm run

Previously it used the global npm, not the npm we were running from.

iarna committed:
extract: Stop reading the package.json here

zkat committed:
pacote: fix typo

zkat committed:
make: stop saving marked-man and prune stuff during build

zkat committed:
pacote: use the current node + npm to run git installs

zkat committed:
feat(shrinkwrap): support extra metadata fields

zkat committed:
shrinkwrap: warn on incompatible lockfile version

zkat committed:
test: fix more tests

zkat committed:
shrinkwrap: update packageIntegrity for spec compliance

zkat committed:
shrinkwrap: save even with warnings

zkat committed:
shrinkwrap: update createdWith field

zkat committed:
shrinkwrap: more changes to package lock fields

zkat committed:
shrinkwrap: even more fixes

zkat committed:
test: disabling lifecycle-signal while coverage issue figured out

zkat committed:
meta: updated package-lock

zkat committed:
test: standard

zkat committed:
shrinkwrap: stop filling in pkginfo.problems

zkat committed:
docs: documented some new installer features

zkat committed:
docs: update package-lock/shrinkwrap documentation

zkat committed:
docs: updated git docs with #semver: support info

zkat committed:
docs: updated npm cache docs

zkat committed:
commands: add affordance for install

zkat committed:
feat(save): fix update and improve experience of moving deps between dep types

iarna committed:
uninstall: When removing direct deps don't remove if also transitive dep

fis-cz commented:
Just a question. why you deprecate the file:?

could not be like the file: will still have the same behavior (copy files to local repo) while link: would create a junction or a symlink?

I would say both of them could be handy in some cases.

zkat commented:
@fis-cz because `file:` on directories was not a well-thought-out feature and has very little actual utility. We repeatedly run into folks surprised by its behavior. It requires extensive maneuvering in order to be used during development, as well.

For folks who want to do the package-up-then-install workflow, `file:` on tarballs is still available, so they can run `npm pack ./my-dir` to create a tarball, then `npm i ./my-dir-tarball.tgz`.

fis-cz commented:
@zkat Thanks for explanation, makes sense to me ;)

iarna committed:
utils/link: Make less race prone

Previously it was gently-rming it's target before making a new target
directory.  gently-rm vacuums, removing empty directories.  This was fine
until you run more than one `link` command concurrently, at which point it's
entirely possible for one invocation to `mkdir` the directory, then
immediately have another one `gently-rm` it, at which point the symlink
fails.  Races!

zkat committed:
pacote@2.7.21

zkat committed:
cacache@9.2.5

zkat committed:
fix(perms): pacote.extract does not run scripts. we already have perm-dropping for cache stuff

iarna committed:
finalize: Make symlinks relative when possible

iarna committed:
fix(shrinkwrap): clarify reuse of sw resolved for diff-registry children

zkat committed:
ssri@4.1.3

iarna committed:
install: Run preinstall prior to reading the tree

iarna committed:
install-deps: Match symlinked transitive deps correctly

zkat committed:
install: call saveRequested even if no changes, to regenerate package-lock

zkat committed:
standard: fix some loose ends

zkat committed:
error-handler: force an exit instead of waiting for tasks

zkat committed:
save: use removedChildren instead of removed

zkat committed:
uninstall: remove stray console.log

iarna committed:
install: Make removing packages work w/o package.json

iarna committed:
install,uninstall: Disable fake children when installing one mod or uninstalling

iarna committed:
test: modernize uninstall-save

iarna committed:
uninstall: defer arg clearing till after global dir is filtered

iarna committed:
test: make no-scan-full-global-dir compat w/ class based install/uninstall

iarna committed:
test: extraneous deps in legacy shrinkwraps are not installed

iarna committed:
install: no need to install non-requested deps

Previously we were doing this so we could load relationships, but now we can
simply request no fake children.  This forces a fetch-metadata phase for the
lockfile, but does not actually install all the modules.

iarna committed:
test: gently-rm-linked-module: modernize test

zkat committed:
pkglock: add --no-package-lock option as --no-shrinkwrap alias

iarna committed:
remove-deps: Update dependencies list even if we're not saving

zkat committed:
pacote: send username/password auth through, too

iarna committed:
test: gently-rm-linked-module: Cleanup output

zkat committed:
pacote: look for always-auth too

iarna committed:
install: allow parseable output for nodes w/o paths

This can happen when, for example, removing a module from your project that
isn't actually in your node_modules at the moment.

iarna committed:
ls: Nothing is extraneous without a package.json

It used to be that we did this check in `is-extraneous` but it turns out to
be much cleaner to keep that pure.  It let's us use `is-extraneous` for
things like pruning.

zkat committed:
safe-buffer@5.0.1

zkat committed:
pacote: decode password before passing it on

zkat committed:
pacote: more alwaysAuth logic

iarna committed:
preinstall: Runs in the final dest, not the staging folder

iarna committed:
glob@7.1.2

iarna committed:
cacache@9.2.5

zkat committed:
lifecycle: added prepack and postpack (#16725)

whitecolor commented:
But there is case where symlinks are inappropriate. Like when you use the same node_modules on a different system like native and docker/windows WSL. Haven't you thought about it?

akosyakov commented:
@zkat How will you set up multiple local packages with common dependencies? We've tried to use `npm link` at the beginning but it has a lot of drawbacks like duplicate dependencies and referenced js/ts files are not resolved properly. Switching to local path dependencies was an improvement since there was no difference in `node_modules` structure in dev and prod mode. We had to use watching though to synchronize changes in dependencies, like [here](https://github.com/theia-ide/theia/blob/master/examples/browser/package.json#L15) for `theia-core`. It sounds to me that with npm5 it won't be possible. What is a proper way?

whitecolor commented:
@akosyakov I currently use [yalc](https://github.com/whitecolor/yalc) as a simple way to work with local packages that are in development phase. It acts like a simple local registry where you "publish" your package and it can automatically propagate changes to all installations, And not symlinks involved, which can be really cumbersome at least for the current state of affairs.

akosyakov commented:
@whitecolor thx, I will give it a try

nicojs commented:
For anyone looking to get the old functionality back: I've created a small package which does this: https://www.npmjs.com/package/install-local. It simply runs an `npm pack` followed by an `npm install <tarball file>`.

@akosyakov @whitecolor I've looked at `yalc` as well, but i've read that it still uses `npm link` under the hood. I'll might give it a shot for future projects. Thanks for sharing.

akosyakov commented:
@nicojs nice

`yalc add` install packaged modules from the local repository. if you do `yalc link` then, yes, you have same issues as with `npm link`.

We've given a try to `yalc`: https://github.com/theia-ide/theia/pull/156#issuecomment-304532963 We don't really have an interest in having the repository for local packages but rather want to set up a local development env for several independent node packages and have fast turnarounds to build them. Right now we are experimenting with [our own tooling](https://github.com/theia-ide/theia/blob/master/config/local-dependency-manager/README.md). For example for [this package](https://github.com/theia-ide/theia/blob/master/examples/browser/package.json) one can run `npm run localinstall` to install declared [local dependencies](https://github.com/theia-ide/theia/blob/master/examples/browser/package.json#L17) above remote and then `npm run watch:localdeps` to start `watch` script on all local dependencies and synchronize changes.

wclr commented:
> I've looked at yalc as well, but i've read that it still uses npm link under the hood.

No, it doesn't use `npm link`, but it allows to work in the same manner, when you don't want to touch your `package.json` (add dependency). But actually `npm link` it is kind of not very consistent fashion of replacing dependencies with other versions. 

[Yalc](https://github.com/whitecolor/yalc) was made to be able consistently to add code of **dependency packages** in your multiple projects. It not for managing dependencies of monorepos or something.  it just adds dependency package from local repository to your project in your node_modules. Consistently means that you doesn't use symlinks, and install all the dependency packages, and I usually keep this code in the project repo and commit it, to stay fully consistent and not to wary about publishing (in npm or git) my dependency packages. Linking capabilities are added more like experiment, I don't use them.

MikeMatusz commented:
As @shadowmint mentioned on #16990, the symlink doesn't honor the `files` section of the `package.json` and this can prove problematic.  #10343 has been around for awhile because you end up with a duplicate node_modules folder.  This change basically makes all of the open issues around symlinks a lot more severe.  I had been avoiding using them specifically because of the issues, and now if I want to use npm@5 I'm forced to deal with them.  I'd love to take advantage of all the improvements in the new version, and I'd even prefer symlinks if they behaved the same as an actual install, but until they do I think this is a deal breaker for my team. :(

Would it be possible to symlink only the files/directories that match the pattern of `files` || !`.npmignore` (i.e. the ones that would be present on an actual install)?  Or are there cases where that would create an insane number of links or be impractical for other reasons?

wesleytodd commented:
I just made my first attempt to run our companies applications with `npm@5` and ran into a few snags with what I think is this feature (was led here by [this post](http://blog.npmjs.org/post/161081169345/v500)).  My two questions are:

1. Local dependencies don't seem to be getting an `npm i` in them, atleast not a full install.  It looks like deps shared in the root project are not being installed in the symlinked package.  The old `npm link` did run a full install, but it appears that now we have to manually run it in each local dep.  Is that correct?  If so, it seems rather weird because an `npm i` in the root project does not result in a runnable application.
2. Is there a built in way to do the "old" mechanism?  In certain circumstances we have code that relies on singletons (I have long since been pushing our team away from the pattern, but alas, never soon enough).  It would appear we either need to stall on `npm@4` or publish these to our internal npme (aka stop using local deps).  Is this the planned way forward, or are there doc somewhere to ease this transition that I did not find?

Also of note: we are in the process of upgrading from `node@4`, so cannot use the `--preserve-symlinks` flag which I believe would give us something closer to the behavior we desire.

nicojs commented:
If anyone is looking for the old functionality of `npm i file:`, please take a look at https://www.npmjs.com/package/install-local. It should work for all versions of npm since version 2. I will use that for all my projects where i want an actual local install, instead of an npm link. 

@wesleytodd might be useful for your usecase as well.

cchamberlain commented:
`npm` 5 has a few issues on Azure app service as it does not (and will likely never) support symbolic links. See https://github.com/projectkudu/KuduSync/issues/33#issuecomment-310811410

In addition to local packages breaking, npm scripts that reference dev dependency installed executables do not operate correctly (`rimraf`, `cross-env`, `webpack` in scripts all fail, unable to find the executable). Guessing this is probably related to a change in `node_modules/.bin` that involves symbolic links? Seems likely there are other cloud services that will suffer similar breakage.

wclr commented:
symlinks are bad. npm don't cares.

donaldpipowitch commented:
Will npm support a feature similar to `yarn`'s workspaces _natively_ in the future? Is there an issue which I could track? I can't find one.

I'd like to see a way to resolve packages locally without symlinks, with flat dependencies and which respects `.gitignore`/`.npmignore`/`"files"`. E.g. so they are installed like a regularly published package or basically like `git`-based packages, but locally.

wclr commented:
@donaldpipowitch 

> I'd like to see a way to resolve packages locally without symlinks, with flat dependencies and which respects .gitignore/.npmignore/"files". E.g. so they are installed like a regularly published package or basically like git-based packages, but locally.

Look at [yalc](https://github.com/whitecolor/yalc)

donaldpipowitch commented:
> Look at yalc

Yeah, thanks. Just wanted to ask if there are plans to do something similar *natively* in npm?

nicojs commented:
> Yeah, thanks. Just wanted to ask if there are plans to do something similar natively in npm?

There is something *kind of* native. Use `npm pack` in the dependency folder and `npm install ../sibling/sibling.0.0.0.tar.gz` in the dependent folder.

This is what [install-local](https://www.npmjs.com/package/install-local) does for you.

wclr commented:
> Yeah, thanks. Just wanted to ask if there are plans to do something similar natively in npm?

I have doubts about `npm`, `yarn` maybe will do something like that but pretty sure not soon, though is really kind of simple task. If we don't have this natively we have to make/use some *hack* before the feature is supported.

donaldpipowitch commented:
Okay, thank you both.

shadowmint commented:
Given that the open issues around this are just being ignored at the moment
(with all the other npm 5 things that need fixing I guess), I wouldn't hold
your breath for any kind of quick resolution.

I recommend you just stick with npm 4 for now (and probably the foreseeable
future).

On Jul 10, 2017 7:57 PM, "Donald Pipowitch" <notifications@github.com>
wrote:

> Okay, thank you both.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/npm/npm/pull/15900#issuecomment-314084475>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAVtsFjPVa6WxhUs1Gi3BVM6OIeIV08Xks5sMhF7gaJpZM4MPBMV>
> .
>

iarna commented:
@donaldpipowitch
> Will npm support a feature similar to yarn's workspaces natively in the future? Is there an issue which I could track? I can't find one.

Yes, no issue as yet, I need to translate their RFC into something that makes sense more broadly (it was rather yarn specific, to the point that it was hard to follow as an implementor w/o deep yarn knowledge). But once I do, we'll track that in new RFC issue. Assuming they haven't strayed from the RFC they shared with us a few months back, I anticipate our implementation being compatible. The idea is that tools like [`lerna`](https://github.com/lerna/lerna) will get to defer more of the thinking about how things install to the package managers.

At its heart it's pretty simple: `workspaces` property of `package.json`, an array of globs. For all folders that match the globs, install their deps when running `npm install` at the top level and cross-link them when they request any version of a module w/ the same name.

That's currently scheduled for _after_ fixing `npm link`'s behavior as it relates to file dependencies.

iarna commented:
The spec here actually landed in e563d619c, so I'm going to close this PR. I encourage folks looking for further discussion around this implementation to open a new issue or comment on an existing one.

donaldpipowitch commented:
Thank you for the explanation!

UnrememberMe commented:
FYI, the current implementation breaks one of the mono-repo scenario as discovered in the Pants issue linked above.

The scenario is:
1) inside of a mono-repo, we created a wrapper for a commonly used tool, say webpack. 
2) include the wrapper module as a dependency on another module and trying to `npm run` the wrapper module to invoke webpack.

This scenario works in npm3 because webpack as a dependency for the wrapper module will be installed in node_module/.bin. For npm5, a symlink will be created for the wrapper module and when the wrapper trying to invoke webpack, webpack will not be on $PATH.

Basel78 commented:
Hello,
I have LibA with ModuleA and LibB with ModuleB that use ModuleA from LibA
also I have Project that use Both LibA & LibB
how best to import LibA in LibB and both in Project if I am using the following versions
LibA & LibB both are angular4
Project is angular-cli
What I want to achieve here is to avoid the relative path of module in my code, 
1- LibA as external to LibB so it wont be imported once once bundle the libraries,
2- LibA & LibB as externals and will only install them in Project.

BTW: I used  symlink first to reference LibA in LibB but was only able to import ModuleA in relative path whivh what i want to avoid it, since that would mean n Project I would have LibA, and LibB with Liba under node_modules with relative path to ModuleA
and on the Project level will have libA & LibB in node_modules with Import to ModuleA & ModuleB in relative path.
now I am using npm pack and npm install file://path/to/module-a/moduleA-1.2.3.tgz inside LibB
but import { ModuleA } from "LibA"; is giving "error cannot find ModuleA"

also I have noticed that after using the npm pack that the referenced LibA appears in LibB-> node_modules folder but the "d.ts & .ts" files is dist folder does not appear anymore, even though it appear in folder in LibA ->dist.

UnrememberMe commented:
to @dotchev,  sorry I did not actively monitor this issue.  To answer your question, in a mono-repo, a reasonable setup for a tool might be:
1) create a wrapper (component A) for a commonly used tool, say eslint / webpack etc and supply some base/common configs, and expose a new command line
a simple example can be 
mylint.sh =>
```
... overlay $1 on top of eslint.base.config if any and generate a new combined config
eslint --config=generated_config
```
2) other projects (component B) within a mono-repo might reference the wrapper instead of the original tool via a file link and call `mylint my-overlay-lint-config`. 

with npm5, eslint will no longer be in .bin for component B, and invoking `mylint.sh` directly will fail to execute if running under component B directory.
We can certainly package component A into a private npm registry.  However, one advantage of mono-repo is codes can reference their dependencies directly without needing external artifact storages.

octogonz commented:
Hi @iarna,

We're having an issue with the "file://" specifier not being handled correctly by NPM 5.  We have an automated tool that generates a **package.json** with references like this:

```json
"dependencies": {
    "yargs": "~4.6.0",
    "z-schema": "~3.18.3",
    "project1": "file:./projects/project1.tgz",
    "project2": "file:./projects/project2.tgz",
```

The "project1.tgz" archive contains a **package.json** file like this:

```json
{
  "name": "project1",
  "version": "0.0.0",
  "private": true,
```

When NPM 4 installs it, the **node_modules/project1/package.json** looks like this, as expected:
```json
  "version": "0.0.0"
```

However, NPM 5 for some reason writes it like this:
```json
  "version": "file:projects/project1.tgz"
```

Was this an intentional design change?  Is it an NPM 5 bug?

It's causing trouble because [read-package-tree](https://www.npmjs.com/package/read-package-tree) chokes on this JSON value. We get an error like this:

```
ERROR: Failed to parse package.json for project1: Invalid version: "file:projects/project1.tgz"
```

octogonz commented:
I'm pretty sure this is an NPM bug.  I've opened #19006

adrian-gierakowski commented:
@iarna what happened to this proposal? has it been rejected? 

btw. looks like something similar has been [implemented in yarn some time ago](https://github.com/yarnpkg/yarn/pull/3359)

zkat commented:
@adrian-gierakowski [it was merged](https://github.com/npm/npm/blob/latest/doc/spec/file-specifiers.md) and implemented as part of [`npm@5.0.0`](https://github.com/npm/npm/releases/tag/v5.0.0)
